modules/apps_api/lib/apps_api/notification_service.rb

```diff
@@ -53,34 +51,46 @@ def parse_event(event)
 
     def create_hash(app_record:, user:, event:)
       {
-        'uuid' => event['uuid'],
-        'app_record' => app_record,
-        'user_email' => user.body['profile']['email'],
-        'options' => {
-          'first_name' => user.body['profile']['firstName'],
-          'application' => app_record ? app_record['name'] : nil,
-          'time' => format_published_time(event['published']),
-          'privacy_policy' => app_record ? app_record['privacy_url'] : nil,
-          'password_reset' => Settings.vanotify.links.password_reset,
-          'connected_applications_link' => Settings.vanotify.links.connected_applications
+        uuid: event['uuid'],
+        app_record: app_record,
+        user_email: user.body['profile']['email'],
+        options: {
+          first_name: user.body['profile']['firstName'],
+          application: app_record ? app_record['name'] : nil,
+          time: format_published_time(event['published']),
+          privacy_policy: app_record ? app_record['privacy_url'] : nil,
+          password_reset: Settings.vanotify.links.password_reset,
+          connected_applications_link: Settings.vanotify.links.connected_applications
         }
       }
     end
 
     def format_published_time(published)
       # Formats iso8601 time stamp into readable language
-      # 2020-11-29T00:23:39.508Z -> 11/29/2020 at 00:23:39:23AM
-      Time.zone.parse(published).strftime('%m/%d/%Y at %T:%M%p')
+      # 2020-11-29T00:23:39.508Z -> 11/29/2020 at 12:23 a.m.
+      Time.zone.parse(published).strftime('%m/%d/%Y at %I:%M %P').sub(/([ap])m/, '\1.m.')
     end
 
-    def event_is_invalid?(event)
+    def event_is_invalid?(parsed_hash, event)
       # checking if the event is unable to be processed,
       # or has already been processed.
-      event_already_handled?(event['uuid']) || event_unsuccessful?(event)
+      event_already_handled?(parsed_hash) || event_unsuccessful?(event)
     end
 
-    def event_already_handled?(uuid)
-      @handled_events.include?(uuid)
+    def event_already_handled?(parsed_hash)
+      already_handled = false
+      # get all members of the notification_events set
+      members = Redis.current.smembers('apps_notification_events')
+      return false if members.nil?
+
+      # check if there is a member with the same email
+      # and time as our parsed_hash to detect duplicates
+      current_event = { 'email' => parsed_hash[:user_email], 'time' => parsed_hash[:options][:time] }
+      members.each do |member|
+        member_hash = Redis.current.hgetall(member)
+        already_handled = true if current_event.eql? member_hash
```

Comment:
can this just return true here? instead of having the `already_handled` variable
i dont see a reason why the rest of the list of members needs to be iterated over if it finds an event that matches the event youre checking

Reply:
This is a good catch, thanks. I rewrote this a couple times and that must have gotten missed.